dmoj/judge.py

```diff
@@ -284,22 +223,238 @@ def listen(self):
+        self.process.start()
+        child_conn.close()
+
+        while True:
+            try:
+                ipc_msg = self.process_conn.recv()
+            except Exception:
+                print("Failed to read message from grading subprocess", file=sys.stderr)
+                raise
+
+            if ipc_msg is None:
+                self.process_conn.send(None)
+                break
+
+            yield ipc_msg
+
+        self.process_conn.close()
+        self.process.join()
+
+    def kill(self):
+        self.process.kill()
+
+    def join(self, timeout=None):
+        self.process.join(timeout=timeout)
+
+    def abort_grading(self):
+        if self.process:
+            try:
+                self.process_conn.send((IPC_REQUEST_ABORT, None))
+            except Exception:
+                print("Failed to send abort request, did it race?", file=sys.stderr)
+                traceback.print_exc()
+            self.join()
+
+    def _do_abort(self):
+        self._abort_requested = True
+        if self.grader:
+            self.grader.abort_grading()
+
+    # TODO(tbrindus): do we need some sort of barrier here? what happens if an abort is sent before
+    # _block_and_grade is called?
+    def _run_judge_request_handler_thread(self, conn):
+        while True:
+            try:
+                ipc_msg = conn.recv()
+            except Exception:
+                print("Parent process unexpectedly hung up!", file=sys.stderr)
+                traceback.print_exc()
+                self._do_abort()
+                return
+
+            if ipc_msg is None:
+                return
+
+            type, data = ipc_msg
+            if type == IPC_REQUEST_ABORT:
+                self._do_abort()
+            else:
+                raise RuntimeError("unexpected IPC message %s" % str(ipc_msg))
+
+    def _run_grading_wrapper_process(self, conn):
+        self.process_conn.close()
+
+        comm_thread = None
+        try:
+            comm_thread = threading.Thread(target=self._run_judge_request_handler_thread, args=(conn,), daemon=True)
+            comm_thread.start()
+
+            for type, data in self._block_and_grade_cases():
+                conn.send((type, data))
+
+            # Signal to the other end that we're done.
+            conn.send(None)
+        except BrokenPipeError:
+            # There's nothing we can do about this... the general except branch would just fail again. Just re-raise and
+            # hope for the best.
+            raise
+        except:  # noqa: E722, we explicitly want to notify the parent of everything
+            # We can't pickle the whole traceback object, so just send the formatted exception.
+            exc = sys.exc_info()
+            message = ''.join(traceback.format_exception(*exc))
+            conn.send((IPC_UNHANDLED_EXCEPTION, (message,)))
+            conn.send(None)
+        finally:
+            if comm_thread is not None:
+                comm_thread.join(timeout=5)
```

Comment:
What happens if this timeout is exceeded?

Reply:
The subprocess exits anyway without necessarily having done the dual-FIN exchange, and the main judge will get an error when trying to read from the pipe. But we can't really wait here forever, I think.

Reply:
Added logging.